storage/src/test/java/zipkin2/storage/kafka/streams/TraceStorageTopologyTest.java

```diff
@@ -144,72 +140,70 @@
-        .localEndpoint(Endpoint.newBuilder().serviceName("svc_a").build())
-        .timestamp(10000L).duration(11L)
-        .putTag("environment", "dev")
-        .build();
+      .localEndpoint(Endpoint.newBuilder().serviceName("svc_a").build())
+      .timestamp(10000L).duration(11L)
+      .putTag("environment", "dev")
+      .build();
     List<Span> spans = Arrays.asList(a, b, c);
-    testDriver.pipeInput(factory.create(spansTopic, a.traceId(), spans, 10L));
+    factory.pipeInput(a.traceId(), spans, 10L);
     // Then: trace stores are filled
-    KeyValueStore<String, List<Span>> traces = testDriver.getKeyValueStore(TRACES_STORE_NAME);
-    assertThat(traces.get(a.traceId())).containsExactlyElementsOf(spans);
-    KeyValueStore<Long, Set<String>> spanIdsByTs =
-        testDriver.getKeyValueStore(SPAN_IDS_BY_TS_STORE_NAME);
-    KeyValueIterator<Long, Set<String>> ids = spanIdsByTs.all();
-    assertThat(ids).hasNext();
-    assertThat(ids.next().value).containsExactly(a.traceId());
-    // Then: service name stores are filled
-    KeyValueStore<String, String> serviceNames =
-        testDriver.getKeyValueStore(SERVICE_NAMES_STORE_NAME);
-    List<String> serviceNameList = new ArrayList<>();
-    serviceNames.all().forEachRemaining(serviceName -> serviceNameList.add(serviceName.value));
-    assertThat(serviceNameList).hasSize(2);
-    assertThat(serviceNames.get("svc_a")).isEqualTo("svc_a");
-    assertThat(serviceNames.get("svc_b")).isEqualTo("svc_b");
-    KeyValueStore<String, Set<String>> spanNames =
-        testDriver.getKeyValueStore(SPAN_NAMES_STORE_NAME);
-    assertThat(spanNames.get("svc_a")).containsExactly("op_a");
-    assertThat(spanNames.get("svc_b")).containsExactly("op_b");
-    KeyValueStore<String, Set<String>> autocompleteTags =
-        testDriver.getKeyValueStore(AUTOCOMPLETE_TAGS_STORE_NAME);
-    assertThat(autocompleteTags.get("environment")).containsExactly("dev");
-    // When: clock moves forward
-    Span d = Span.newBuilder()
-        .traceId("d")
-        .id("d")
-        .timestamp(
-            MILLISECONDS.toMicros(traceTtlCheckInterval.toMillis()) + MILLISECONDS.toMicros(20))
-        .build();
-    testDriver.pipeInput(
-        factory.create(spansTopic, d.traceId(), Collections.singletonList(d),
-            traceTtlCheckInterval.plusMillis(1).toMillis()));
-    // Then: Traces store is empty
-    assertThat(traces.get(a.traceId())).isNull();
+    WindowStore<String, List<Span>> traces = testDriver.getWindowStore(TRACES_STORE_NAME);
+    try (final WindowStoreIterator<List<Span>> fetch = traces.fetch(a.traceId(), 0, 10000L)) {
+      assertThat(fetch.hasNext()).isTrue();
+      final KeyValue<Long, List<Span>> next = fetch.next();
+      assertThat(next.value).isEqualTo(spans);
+    }
+    WindowStore<String, Set<String>> spanNames = testDriver.getWindowStore(SPAN_NAMES_STORE_NAME);
+    Map<String, Set<String>> serviceNames = new HashMap<>();
```

Comment:
by mentioning using `.extracting` I meant to see if you can avoid copying into a map first. ex assert on the WindowStore.fetchAll

not a biggie and I can give a try if you like (tomorrow)

Reply:
Right! Got it. Hope it looks better now 😅